yt_dlp/extractor/zenporn.py

```diff
@@ -0,0 +1,151 @@
+            'post_date': '20231024',
+            'uploader': 'Martin Rudenko',
+            'age_limit': 18
+        }
+    }]
+
+    def _gen_info_url(self, ext_domain, extr_id, lifetime=86400):
+        """ This function is a reverse engineering from the website javascript """
+        extr_id = int_or_none(extr_id)
+        if extr_id is None:
+            raise ExtractorError('Unable to generate the `gen_info_url`')
+        result = '/'.join(str(extr_id // i * i) for i in (1_000_000, 1_000, 1))
+        return f'https://{ext_domain}/api/json/video/{lifetime}/{result}.json'
+
+    def _decode_video_url(self, ext_domain, encoded_url):
+        """ This function is a reverse engineering from the website javascript """
+        cust_char_set = 'АВСDЕFGHIJKLМNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789.,~'
+        decoded_url = ''
+        cur_pos = 0
+
+        # Check for characters not in the custom character set and handle errors
+        if any(char not in cust_char_set for char in encoded_url):
+            return None
+
+        # Filter out characters not in the custom character set
+        encoded_url = ''.join(char for char in encoded_url if char in cust_char_set)
+
+        while cur_pos < len(encoded_url):
+            o = cust_char_set.index(encoded_url[cur_pos])
+            i = cust_char_set.index(encoded_url[cur_pos + 1])
+            s = cust_char_set.index(encoded_url[cur_pos + 2])
+            a = cust_char_set.index(encoded_url[cur_pos + 3])
+
+            o = (o << 2) | (i >> 4)
+            i = ((i & 15) << 4) | (s >> 2)
+            l = ((s & 3) << 6) | a
+
+            decoded_url += chr(o)
+            if s != 64:
+                decoded_url += chr(i)
+            if a != 64:
+                decoded_url += chr(l)
+
+            cur_pos += 4
+
+        return f'https://{ext_domain}{decoded_url}'
+
+    def _real_extract(self, url):
+
+        video_id = self._match_id(url)
+        webpage = self._download_webpage(url, video_id)
+
+        ext_domain, extr_id = self._search_regex(
+            r'https:\/\/(?P<ext_domain>[\w.-]+\.\w{3})\/embed\/(?P<extr_id>\d+)\/',
+            webpage, 'embed_info', group=('ext_domain', 'extr_id'))
+
+        info_json = self._download_json(self._gen_info_url(ext_domain, extr_id),
+                                        video_id, note="Downloading JSON metadata for the video info.")
+        if not info_json:
+            raise ExtractorError('Unable to retrieve the video info.')
+
+        video_json = self._download_json(f'https://{ext_domain}/api/videofile.php?video_id={extr_id}&lifetime=8640000',
+                                         video_id, note="Downloading JSON metadata for the video location.")
+        if not video_json:
+            raise ExtractorError('Unable to retrieve the the video location.')
+
+        encoded_url = video_json[0].get('video_url')
+        if not encoded_url:
+            raise ExtractorError('Unable to retrieve the `encoded_url` value.')
```

Comment:
too many `ExtractorError`s that say basically the same thing a `KeyError` would
```suggestion
        encoded_url = video_json[0]['video_url']
```